### ISSUE TITLE
[Superseded]: Colocate feature code where useful; do not add CapabilitySlice manifests or registries

### ISSUE THREAD

**Issue by morluto** [morluto/jacobian#1056]
## Disposition

Closed as not planned. Superseded by #1491, #1222’s dependency-direction work, and concrete domain cleanup issues.

This issue correctly identified that unrelated capability/checker PRs collide in broad contract, operation, checker, registry, and portfolio-test hubs.

Its proposed replacement—typed `CapabilitySlice`, per-feature manifests, generated slice registries, checker/provider/protocol declarations, slice-graph validation, compiled installation plans, registry-generated docs/tests, and migration/dependency declarations—would create another internal plugin/feature framework.

## Simpler endpoint

Use ordinary Python modules and imports:

- place feature-specific values/functions/tests together when they change together;
- keep genuinely shared mathematical values in one lower-level owner;
- split broad checker/operation/test files when concrete independent reasons to change are demonstrated;
- allow one simple explicit import/tuple/dict aggregation point if current package discovery needs it;
- remove operation-specific branches from central hubs where direct imports/functions suffice.

Do not require every feature to instantiate a common slice object or manifest.

## Retained work

- split high-conflict number-theory/polynomial/graph modules along concrete mathematical ownership;
- move independent checker implementations out of cross-domain monoliths;
- keep producer/checker implementations independent;
- keep operation-specific bounds beside the operation rather than a broad shared value;
- replace central bespoke test expectation lists with ordinary parametrization over existing static entries where useful;
- delete stale central maps/branches as features move.

## Do not introduce

- `CapabilitySlice` or feature manifest objects;
- a slice registry/generator/compiler;
- per-slice provider/checker/protocol/doc/example declarations;
- slice graph dependency/migration validation;
- generated installation plans or atomic slice publication;
- a new internal plugin system or package-template mandate;
- one directory per trivial feature.

A small module-level tuple/dict is acceptable if it is the simplest existing aggregation mechanism and contains no executable framework semantics.

## Canonical owners

- #1491 — one semantic owner per value/feature without prescribed layout.
- #1222 — enforce dependency direction, not a feature framework.
- #1478/#1479/#1489/#1498 — concrete package/domain consolidation.
- #1484/#1495 — direct checker ownership without slice protocols.

**Comment by morluto** [comment]
## Mature-library repository-organization addendum — vertical slices inside dependency rings

The external comparison supports the vertical-slice direction, but adds an important qualification: mature libraries localize implementation changes without allowing one feature module to depend on every architectural layer.

### PyTorch: central semantic definitions, distributed implementations

PyTorch keeps the semantic operator definition centralized in a declarative registry and permits per-backend implementations/registrations to live in separate files or even separate libraries.

The common structure documented by the dispatcher tutorial is:

```text
one namespace/operator definition surface
+ implementation registration per dispatch key
+ implementations colocated with the relevant backend/operator code
```

- https://docs.pytorch.org/tutorials/advanced/dispatcher
- https://docs.pytorch.org/cppdocs/api/library/registration.html
- https://github.com/pytorch/pytorch/blob/main/aten/src/ATen/native/README.md

`torchgen` then turns the semantic declaration into generated boundary artifacts rather than asking every backend file to rebuild the public API.

#### Revision for Jacobian

A `CapabilitySlice` should own the feature's semantic declaration and references to implementation/checker bindings, but it should **not** manually construct schemas, adapters, result envelopes, persistence, or central registry mutations.

A useful split is:

```text
feature values/contracts
feature pure kernel(s)
feature operation/protocol declaration
provider implementation binding(s)
independent checker contract/entrypoint
focused tests
```

The common compiler generates:

```text
CapabilityDescriptor
MCP/native selected-call schemas
adapter/dispatch binding
catalog relationships
artifact publication plan
docs/inventory projections
aggregate conformance cases
```

This preserves local change ownership while preventing each slice from becoming a miniature runtime.

### JAX: make public, extension, private, and experimental surfaces explicit

JAX documents which APIs are public, treats underscore-prefixed and `jax._src` paths as private, and exposes `jax.extend` as a deliberate lower-stability integration surface rather than making downstream projects import arbitrary internals.

- https://docs.jax.dev/en/latest/api_compatibility.html
- https://docs.jax.dev/en/latest/jep/15856-jex.html

#### Revision for Jacobian

The slice architecture should produce three explicit inventories:

```text
public mathematical/API exports
supported extension declarations/protocols
private runtime/compiler implementation
```

A slice must not make its internal installer or adapter the extension API merely because that class is importable.

Coordinate with #1222 for the repository-wide package DAG and extension namespace.

### Arrow: keep value definitions, kernels, and transport separate

Arrow keeps logical types/schemas, concrete compute kernels, and IPC/in-process transport distinct. A feature can add a logical extension type and kernels without redefining the transport protocol.

- https://arrow.apache.org/docs/format/Intro.html
- https://arrow.apache.org/docs/cpp/api/compute.html

#### Revision for Jacobian

Feature slices should reference shared mathematical value/parent/axis contracts from #1057/#1129/#1140 and typed ports from #1208. They should not create operation-specific wrappers solely because a value is inline versus durable.

Persistence is a runtime policy over a value, not the feature's type identity.

### MLIR: explicit static registration and generic interfaces

MLIR uses declarative operation definitions and explicit interfaces so generic infrastructure does not require edits for every dialect operation.

- https://mlir.llvm.org/docs/DefiningDialects/Operations/
- https://mlir.llvm.org/docs/Interfaces/

#### Revision for Jacobian

The static slice registry should contain passive immutable declarations. Registration should be generated/compiled, not performed by import-time side effects.

Adding a feature should not require operation-specific branches in:

```text
checker runtime selection
support-predicate maps
MCP projection maps
adapter selection
result-envelope construction
architecture allowlists
```

If one of those generic systems needs a new semantic concept, add one typed interface/policy implemented by several features—not one ID-specific branch.

## Proposed package shape

The directory layout may remain domain-oriented, but each slice should obey dependency rings from #1222. For example:

```text
src/jacobian/domains/number_theory/finite_field_map/
  values.py                 # parent/element/map/certificate contracts
  kernel.py                 # pure producer mathematics
  operation.py              # OperationDefinition + ports + policies
  provider_flint.py         # provider binding
  checker_contract.py       # passive verification protocol declaration

src/jacobian_checkers/number_theory/finite_field_map.py
  # independent executable replay; no producer-kernel import
```

A generated static inventory can refer to the operation and checker declarations. The feature does not call the compiler or mutate the registry itself.

## Acceptance additions

- [ ] A slice is passive at import time: importing it performs no registration, provider probing, storage access, process spawn, or global mutation.
- [ ] A slice exposes one canonical operation/protocol declaration consumed by the common compiler; it does not construct `CapabilityDescriptor`, `CapabilityResult`, or MCP bindings directly.
- [ ] Backend/provider implementations may live in separate modules/packages and bind to the same semantic operation definition without duplicating schemas.
- [ ] The generated registry is the only aggregate edit, or is regenerated automatically; domain bundles and docs do not maintain separate hand-authored inventories.
- [ ] Public/extension/private exports are generated or checked from the slice inventory.
- [ ] Feature behavior tests live with the slice; aggregate tests iterate generic invariants rather than append one bespoke assertion per ID.
- [ ] Removing one slice fails if another slice depends on its value/port contract without an explicit declared dependency.
- [ ] A slice can be moved between inline/durable providers or gain another provider without reorganizing its mathematical contracts.
- [ ] The package DAG from #1222 prevents a feature-owned module from importing runtime/MCP/storage simply to remain vertically colocated.

## Important non-goal

Do not interpret “vertical slice” as “put every layer for one feature into one Python module.” The unit of ownership may be one directory/package with inward-pointing submodules. File adjacency is useful; dependency collapse is not.

**Comment by morluto** [comment]
## Library-organization research: combine stable horizontal foundations with volatile vertical slices

The mature repositories do **not** use one universal package rule. Their useful commonality is that package boundaries follow ownership/stability and dependency direction, while extension glue is generated or mechanically registered.

### PyTorch's layered foundation

PyTorch explicitly separates:

```text
c10       essential dependency-light core usable in constrained builds
ATen      C++ tensor/operator library without autograd
native    operation implementations, often separated by backend
 torch     Python frontend and C++ application/binding code
 torchgen  generation from authoritative operation declarations
```

Source:

- https://github.com/pytorch/pytorch/blob/main/CONTRIBUTING.md#codebase-structure

This is deliberately horizontal because these layers have different dependency and binary/runtime responsibilities. At the extension seam, however, one operator schema generates bindings and backend implementations register independently against it. A new kernel does not duplicate the Python/C++ public API or edit every frontend manually.

### scikit-learn's domain modules over one protocol

scikit-learn organizes estimators primarily by user-facing mathematical/ML domain (`linear_model`, `cluster`, `decomposition`, etc.) while shared composition depends on a narrow estimator protocol (`fit`, `transform`, `predict`, `get_params`, tags, metadata routing). Generic estimator checks exercise implementations through that common contract.

Sources:

- https://scikit-learn.org/stable/developers/develop.html
- https://scikit-learn.org/stable/modules/compose.html
- https://scikit-learn.org/stable/metadata_routing.html

This is close to Jacobian's desired user-facing domain organization: operations remain discoverable by mathematical domain, while the extension protocol is shared.

### xarray/Arrow: stable value substrate below feature APIs

xarray builds `DataArray` and `Dataset` from a smaller `Variable` value core. Apache Arrow separates logical types/schemas from storage buffers and provides a small stable interchange boundary. Feature APIs compose the stable value substrate rather than duplicating representations.

Sources:

- https://docs.xarray.dev/en/stable/internals/internal-design.html
- https://arrow.apache.org/docs/format/Columnar.html
- https://arrow.apache.org/docs/format/CDataInterface.html

### Revised package rule for Jacobian

Use both axes deliberately:

```text
Horizontal, stable, dependency-light foundations
    contracts/exact values
    domain/parent/axis/value-type protocols
    operation/port/outcome interfaces
    canonicalization
    provider/checker/application SPI

Vertical, independently changing feature slices
    operation request/result types
    operation definition
    abstract evaluator (#1223)
    postcondition
    examples/docs
    provider implementation registration(s)
    verification protocol + independent checker entrypoint
    focused tests

Generated/mechanical aggregation
    domain inventory
    provider/checker registration indexes
    catalog/docs projections
    generic conformance matrices
```

A feature should not duplicate stable foundations to achieve locality. Conversely, feature-specific semantics should not be appended to broad `contracts/<domain>.py`, `operations.py`, checker maps, or portfolio tests merely because the feature belongs to that domain.

### Separate semantic ownership from deployment packaging

A slice may own the declaration of a provider requirement and checker protocol without packaging producer and checker executable code together.

Keep distinct:

```text
semantic feature slice
    passive operation/type/protocol declarations

producer implementation component
    provider kernel and dependency identity

checker component
    separately packaged clean-process implementation

runtime assembly
    operator-selected installation graph
```

Co-location in the source tree is not authority. Separate executable packaging and import boundaries still enforce checker independence.

### Provider/backend subpackages should be implementation-owned

PyTorch's native CPU/CUDA/MPS/backend folders are useful when implementations have different build/runtime dependencies. Apply this only where demonstrated:

```text
feature definition owns operation semantics
provider implementation owns backend imports/code
```

Do not put the operation definition under a SymPy/Sage/FLINT provider package, and do not make a feature slice import every optional provider.

For a capability with one small built-in implementation, co-locating the typed kernel with the slice is fine. Extract a provider component when it has an independent dependency, lifecycle, runtime identity, or release reason.

### Static registration, lazy optional runtime imports

PyTorch's native-DSL rules prohibit loading optional DSL runtimes or initializing CUDA during registration/import. Heavy kernels are imported lazily when the registered implementation is actually selected.

Source:

- https://github.com/pytorch/pytorch/blob/main/torch/_native/README.md

Apply this to slice assembly:

- registry generation/import must not initialize Sage, Lean, SAT/SMT solvers, GPU runtimes, or external CAS providers;
- provider identity/readiness probing must be bounded and side-effect-safe;
- implementation import happens after policy/provider/admission selection;
- missing optional implementation does not remove or mutate the passive operation definition;
- generated inventory is deterministic and import-order-independent.

### Avoid a `CapabilitySlice` god manifest

`CapabilitySlice` should be an ownership/assembly unit, not another bag containing copied schemas, providers, effects, ports, docs, and checker data.

Prefer references to authoritative declarations:

```python
CapabilitySlice(
    slice_id=...,
    operations=(OperationDefinitionRef(...), ...),
    implementation_registrations=(...),
    verification_protocols=(...),
    docs=(...),
)
```

The compiler resolves and cross-validates these declarations. It should reject duplicated facts rather than accepting a manifest that becomes a second source of truth.

### Common tests should be generated; semantic tests stay local

Borrow the scikit-learn/PyTorch pattern:

```text
feature-local tests
    mathematical examples, boundaries, corrupted evidence, checker independence

registry-generated common checks
    schema/signature, port compatibility, abstract/concrete agreement,
    effects, provider registration, result algebra, contract identity
```

Adding a feature should usually add local fixtures and automatically enter common conformance. It should not append bespoke inventory assertions to a giant central test file.

### Acceptance additions

- [ ] Stable value/operation/protocol foundations are dependency-light horizontal packages; feature semantics are vertical slices over them.
- [ ] Domain user organization is preserved without using domain-wide files as the change unit.
- [ ] Operation definitions are provider-independent; backend implementations may live in provider-owned modules and register against the definition.
- [ ] Producer and checker executable packages remain independent even when one passive feature slice declares their protocol relationship.
- [ ] Registry/catalog/docs/common-test aggregation is generated or mechanically additive and import-order-independent.
- [ ] Catalog construction and registration do not import or initialize optional heavy providers.
- [ ] `CapabilitySlice` references authoritative declarations rather than copying their schemas/metadata.
- [ ] Feature-local semantic tests and generated interface conformance replace central per-feature test accumulation.
- [ ] Removal of one slice deterministically identifies downstream typed dependencies and leaves no central checker/provider cases behind.

The recommended organization is therefore not "make Jacobian look like PyTorch's directory tree." It is: **stable foundations by dependency, features by reason to change, implementations by runtime owner, and generated glue at the seams**.

**Comment by morluto** [comment]
## Checker slices need per-checker implementation manifests, not package-wide source identity

#1228 proves a current blocker to independently versioned slices:

```python
package_source_digest(entrypoint)
    -> validates selected module exists
    -> discards module/function identity
    -> hashes every regular file under the top-level package
```

All `jacobian_checkers.*` entrypoints therefore share one package-source digest. `checker_source_digest()` additionally hashes the entire top-level `jacobian` package because the worker entrypoint is `jacobian.checker_worker:main`.

So even after declarations move into vertical slices:

- modifying checker slice B changes checker slice A's executable digest;
- modifying an unrelated producer/MCP/CLI/storage file changes every checker digest;
- every authorization is coupled to the complete repository package layout.

A slice should contribute an explicit implementation manifest:

```text
entrypoint/function
exact implementation/helper/data materials
passive checker contract bundle digest
narrow worker-runtime digest
resolved provider/dependency identities
```

The static slice registry/compiler validates those materials against the built distribution. Runtime authorization remeasures the declared artifact; it must not infer dependency closure from the top-level package directory.

Add a non-interference gate to this issue:

```text
change unrelated slice B -> slice A checker identity unchanged
change shared declared checker SDK -> all dependent slices change
change producer/MCP/CLI -> checker identities unchanged
```

This is necessary for the Common Closure Principle this issue is applying: feature slices are not independent release/change units while their trust identity hashes every sibling.

**Comment by morluto** [comment]
Bounded implementation follow-up from the 2026-08-13 conflict-hotspot audit: #1405 owns completion and deletion of the legacy exact-checker compatibility seam.

PR #1299 establishes the right direction for new declarations by allowing declaration-owned provider runtimes, but deliberately preserves `_ENTRYPOINT_PROVIDER_RUNTIME_KEYS`, `_provider_runtime_key()` cases, the central runtime-factory matrix, provider-key regrouping, and `_checker_supports(operation_id, payload)` for existing families. Without a finite migration owner, Jacobian would retain two installation/semantics authorities indefinitely.

#1405 inventories and migrates every legacy family, ratchets against new central cases, derives runtime grouping/support policy from declarations, preserves generic installer authorization, and closes only when the module-name/operation-ID semantic maps are deleted. Graph, Smith/matrix, and polynomial conflict clusters are the priority pilots.
